pkg/packagekit/updates.jsx

```diff
@@ -442,21 +447,29 @@ class OsUpdates extends React.Component {
     }
 
     watchUpdates(transProxy) {
-        this.setState({state: "applying", applyTransaction: transProxy});
+        this.setState({state: "applying", applyTransaction: transProxy, allowCancel: transProxy.AllowCancel});
 
         transProxy.addEventListener("ErrorCode", (event, code, details) => this.state.errorMessages.push(details));
+
         transProxy.addEventListener("Finished", (event, exit) => {
-            if (exit == PK_EXIT_ENUM_SUCCESS) {
+            this.setState({applyTransaction: null, allowCancel: null});
+
+            if (exit == PK_EXIT_ENUM_SUCCESS || exit == PK_EXIT_ENUM_CANCELLED) {
                 this.setState({state: "loading", haveSecurity: false, loadPercent: null});
                 this.loadUpdates();
             } else {
                 // normally we get FAILED here with ErrorCodes; handle unexpected errors to allow for some debugging
                 if (exit != PK_EXIT_ENUM_FAILED)
-                    this.state.errorMessages.push(cockpit.format(_("PackageKit reported error code {0}"), exit));
+                    this.state.errorMessages.push(cockpit.format(_("PackageKit reported error code $0"), exit));
                 this.setState({state: "updateError"});
             }
         });
 
+        transProxy.addEventListener("changed", (event, data) => {
+            if ("AllowCancel" in data && data.AllowCancel != this.state.allowCancel)
```

Comment:
I often do this myself, but I think the check `data.AllowCancel != this.state.allowCancel` is superfluous.
React tests for changes anyway, calling `setState` shouldn't trigger any extra work by itself.